src/tsdl_stubs.c

```diff
@@ -17,11 +18,12 @@
  */
 void tsdl_nop (void) { SDL_WasInit(0); return; }
 #else
+#define _DLLAPI extern
```

Comment:
Why does the function `caml_tsdl_log1arg` get declared as `extern` on non windows platform. Isn't `extern` suppose to declare a function prototype whose implementation is not in the file ? In general I have a few other C stubs that do not have such declarations, does it means that they are all broken on windows ?

Reply:
The use of `extern` is canonical for visibility with GCC, even outside of function prototype declarations: https://gcc.gnu.org/wiki/Visibility. That page is mostly about C++ but they also have C examples.

> In general I have a few other C stubs that do not have such declarations, does it means that they are all broken on windows ?

Can you point to one or two examples so I can try to build another example and see if there is a problem?

Reply:
> The use of `extern` is canonical for visibility with GCC, even outside of function prototype declarations: https://gcc.gnu.org/wiki/Visibility. That page is mostly about C++ but they also have C examples.

But it doesn't seem to tell me I need to put `extern` keywords to have stuff exported. I rather not add stuff if it works as is and in my experience it does on Unix platforms. 

> Can you point to one or two examples so I can try to build another example and see if there is a problem?

A small sample.

https://github.com/dbuenzli/ttweetnacl/blob/master/src/ttweetnacl_stubs.c
https://github.com/b0-system/b0/blob/master/src/std/b0_stubs_cpu.c
https://github.com/b0-system/b0/blob/master/src/std/b0_stubs_hash.c
https://github.com/b0-system/b0/blob/master/src/std/b0_stubs_realpath.c
https://github.com/b0-system/b0/blob/master/src/std/b0_stubs_realpath.c
https://github.com/dbuenzli/rel/blob/master/src/rel_sqlite3_stubs.c
https://github.com/dbuenzli/bytesrw/blob/main/src/zstd/bytesrw_zstd_stubs.c

Note that nowaday `CAMLprim` is [empty](https://github.com/ocaml/ocaml/blob/de093ee9601df160060517148e2ae1468d6d6395/runtime/caml/misc.h#L129).

Reply:
My muscle memory tells me I had to stick a bunch of externs when cross-compiling for Android + JNI (yes, unusual) ... and it might be due to the flags that Android NDK uses (see -Bsymbolic at https://android.googlesource.com/platform/ndk/+/HEAD/docs/ClangMigration.md for an earlier slightly related issue). But it is not worth hunting down the exact issue; I'll remove the extern from Unix.

Reply:
After studying rel (have not built it yet) I see that it uses the conventional `external` keyword which, on Windows, is resolved through FlexDLL. FlexDLL advertises itself as not needing declspec on its README ... it uses its own custom symbol table. So it should not need declspec (neither should anything in ocaml/ocaml).

In contrast ctypes uses libffi which is a (somewhat) "normal" user of DLLs ... the symbols have to be available to the Windows LoadLibrary method, so declspec is necessary so that MSVC can embed the symbols in the DLL symbol tables.

So can't draw any conclusions comparing rel and tsdl.